### ISSUE TITLE
FittedTextLayoutCollection - layouts.first can be nil

### ISSUE THREAD

**Issue by PragTob** [shoes/shoes4#636]
I had to add the following guard statement:

```ruby
        return [] unless @layouts.first # TODO WTF
```

to the code for simple-downloader.rb to work. See [this comment](https://github.com/shoes/shoes4/pull/612#issuecomment-39989776).

It seems to work for @jasonrclark and @KCErb though. So maybe it's linux only...

I'd love for more people to try it out though.

Steps to reproduce is basically:

1. start samples/simple-downloader.rb
2. Enter like any URL
3. hit go
4. BOOOOOM

Help/Investigation appreciated.

**Comment by jasonrclark** [comment]
So I don't know that it was the case back when this was written, but from the work that I've done on #1024, it *is* actually possible now for the `@segments` in the collection to end up empty under certain resizing scenarios.

Now I still don't know why you were hitting it during the simple-downloader, but it actually seems prudent to me to keep that guard now. What do you think @PragTob?

**Comment by jasonrclark** [comment]
(Oh, and for anyone following along the `FittedTextLayoutCollection` class was renamed to `TextSegmentCollection` and the `@layouts` instance variable on it is now `@segments` instead (https://github.com/shoes/shoes4/blob/5032b0631bb60f5f96907477d4336316f5f44ea8/shoes-swt/lib/shoes/swt/text_block/fitter.rb)

**Comment by PragTob** [comment]
Hey @jasonrclark !

Unfortunately I still hit this bug on both master and on your #1024 branch.... :(

```
tobi@speedy ~/github/shoes4 $ bin/shoes samples/simple-downloader.rb 
NoMethodError: undefined method `text' for nil:NilClass
        segment_ranges at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb:88
  style_segment_ranges at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb:43
                  each at org/jruby/RubyHash.java:1341
  style_segment_ranges at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb:41
         paint_control at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb:18
          draw_layouts at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/painter.rb:26
          paintControl at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/painter.rb:19
          paintControl at Shoes$$Swt$$TextBlock$$Painter_251182584.gen:13
            initialize at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/progress.rb:16
                  call at org/jruby/RubyMethod.java:132
           backend_for at /home/tobi/github/shoes4/shoes-core/lib/shoes/configuration.rb:57
            initialize at /home/tobi/github/shoes4/shoes-core/lib/shoes/progress.rb:16
                create at /home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:193
              progress at /home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:223
                (root) at /home/tobi/github/shoes4/samples/simple-downloader.rb:15
                  call at org/jruby/RubyProc.java:271
            eval_block at /home/tobi/github/shoes4/shoes-core/lib/shoes/slot.rb:52
            initialize at /home/tobi/github/shoes4/shoes-core/lib/shoes/slot.rb:20
                create at /home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:193
                 stack at /home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:245
                (root) at /home/tobi/github/shoes4/samples/simple-downloader.rb:11
                  call at org/jruby/RubyProc.java:271
            eval_block at /home/tobi/github/shoes4/shoes-core/lib/shoes/slot.rb:52
            initialize at /home/tobi/github/shoes4/shoes-core/lib/shoes/slot.rb:20
                create at /home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:193
                 stack at /home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:245
                (root) at /home/tobi/github/shoes4/samples/simple-downloader.rb:9
                  call at org/jruby/RubyProc.java:271
            eval_block at /home/tobi/github/shoes4/shoes-core/lib/shoes/slot.rb:52
                append at /home/tobi/github/shoes4/shoes-core/lib/shoes/slot.rb:71
                (root) at /home/tobi/github/shoes4/samples/simple-downloader.rb:8
                  call at org/jruby/RubyProc.java:271
            eval_block at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/swt_button.rb:24
            eval_block at /home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/after_do-0.3.1/lib/after_do.rb:99
                 click at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/swt_button.rb:33
        widgetSelected at org/jruby/gen/InterfaceImpl324678930.gen:13
        run_event_loop at /home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/swt-4.4/lib/swt/minimal.rb:109
            event_loop at /home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/swt-4.4/lib/swt/minimal.rb:92
                  open at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/app.rb:45
              open_gui at /home/tobi/github/shoes4/shoes-core/lib/shoes/internal_app.rb:108
            initialize at /home/tobi/github/shoes4/shoes-core/lib/shoes/app.rb:49
                   app at /home/tobi/github/shoes4/shoes-core/lib/shoes/app.rb:19
                (root) at /home/tobi/github/shoes4/samples/simple-downloader.rb:1
                  load at org/jruby/RubyKernel.java:1087
                (root) at /home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:1
           execute_app at /home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:49
                   run at /home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:63
                  load at org/jruby/RubyKernel.java:1087
                (root) at /home/tobi/github/shoes4/bin/shoes-swt:5
```

**Comment by jasonrclark** [comment]
Oh, I misunderstood and thought that you were wanting to remove that hack that you'd used to make it work... didn't realize it was still busted. Although I don't dup it, will look deeper with that stack trace since I was just in the area. Maybe I can turn something up.

**Comment by PragTob** [comment]
I do want to remove the hack - because it's a hack and I don't like hacks :D Maybe the hack is the "solution" but I at least want to know the root cause as it could point to a larger problem.

**Comment by jasonrclark** [comment]
I hear you. Unless you want to debug into it yourself, since I can't dup it, was thinking to get you a diagnostic branch that'll log what's happening at a number of key points, which might help me understand it better. How's that sound?

**Comment by PragTob** [comment]
I think I'll take some time this weekend to debug it, if I don't ping me here, tell me I should get to work and give me that branch :smile:

**Comment by PragTob** [comment]
First results/hunch.. it doesn't seem like `TextBlock#contents_alignment` ever runs, hence the `@segments` is empty... when we want to paint the TextBlock :(

**Comment by PragTob** [comment]
The most minimal example I could get to:

```ruby
Shoes.app do
  button "BAM" do
    para link("Link + Progress = BADABAM")
    progress
  end
end

```

**Comment by jasonrclark** [comment]
Nicely done getting that down to minimal!

Unfortunately, it still doesn't fail for me locally, but I have confirmed part of your suspicion--the ordering does appear to be a `paint` before `contents_alignment` (triggered by something in SWT when we write to the progress control). For me, though, the guards prevent that early paint from failing.

I've pushed up a branch, `diag-636` with some puts statements at key locations--when we paint the text, when we contents align, and some variable values when enter `segment_ranges`. If you'd be so kind as to run that @PragTob, maybe we can shed a bit of light on what's going on there.

**Comment by PragTob** [comment]
Oh I was still on that running with my own set of `puts` statements, just went to bed :) Thanks, will check out your branch and see what happens :)

**Comment by PragTob** [comment]
FYI output from the branch is the following... interesting that the new SWT progressBar triggers a redraw :(

Will look into this further:

```
tobi@tobi-desktop ~/github/shoes4 $ bin/shoes testing/para_fu.rb 

TextBlock::Painter#paintControl (Shoes::Para:0x000039a4)
Shoes$$Swt$$TextBlock$$Painter_1611367039.gen:13:in `paintControl'
	/home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/progress.rb:16:in `initialize'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/configuration.rb:57:in `call'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/configuration.rb:57:in `backend_for'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/progress.rb:16:in `initialize'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:193:in `create'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:223:in `progress'
	/home/tobi/github/shoes4/testing/para_fu.rb:4:in `(root)'
	/home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/swt_button.rb:24:in `call'
	/home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/swt_button.rb:24:in `eval_block'
	/home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/after_do-0.3.1/lib/after_do.rb:99:in `eval_block'
	/home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/swt_button.rb:33:in `click'
	org/jruby/gen/InterfaceImpl1903212855.gen:13:in `widgetSelected'
	/home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/swt-4.4/lib/swt/minimal.rb:109:in `run_event_loop'
	/home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/swt-4.4/lib/swt/minimal.rb:92:in `event_loop'
	/home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/app.rb:45:in `open'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/internal_app.rb:108:in `open_gui'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/app.rb:49:in `initialize'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/app.rb:19:in `app'
	/home/tobi/github/shoes4/testing/para_fu.rb:1:in `(root)'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:1:in `load'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:1:in `(root)'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:49:in `execute_app'
	/home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:63:in `run'
	/home/tobi/github/shoes4/bin/shoes-swt:5:in `load'
	/home/tobi/github/shoes4/bin/shoes-swt:5:in `(root)'

TextSegmentCollection#segment_ranges
[]
{0..24=>[(Shoes::Link:0x000039a8 "Link + Progress = BADABAM")]}
0..24
NoMethodError: undefined method `text' for nil:NilClass
        segment_ranges at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb:92
  style_segment_ranges at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb:43
                  each at org/jruby/RubyHash.java:1341
  style_segment_ranges at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb:41
         paint_control at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb:18
          draw_layouts at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/painter.rb:29
          paintControl at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/text_block/painter.rb:22
          paintControl at Shoes$$Swt$$TextBlock$$Painter_1611367039.gen:13
            initialize at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/progress.rb:16
                  call at org/jruby/RubyMethod.java:132
           backend_for at /home/tobi/github/shoes4/shoes-core/lib/shoes/configuration.rb:57
            initialize at /home/tobi/github/shoes4/shoes-core/lib/shoes/progress.rb:16
                create at /home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:193
              progress at /home/tobi/github/shoes4/shoes-core/lib/shoes/dsl.rb:223
                (root) at /home/tobi/github/shoes4/testing/para_fu.rb:4
                  call at org/jruby/RubyProc.java:271
            eval_block at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/swt_button.rb:24
            eval_block at /home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/after_do-0.3.1/lib/after_do.rb:99
                 click at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/swt_button.rb:33
        widgetSelected at org/jruby/gen/InterfaceImpl1903212855.gen:13
        run_event_loop at /home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/swt-4.4/lib/swt/minimal.rb:109
            event_loop at /home/tobi/.rvm/gems/jruby-1.7.18@shoes/gems/swt-4.4/lib/swt/minimal.rb:92
                  open at /home/tobi/github/shoes4/shoes-swt/lib/shoes/swt/app.rb:45
              open_gui at /home/tobi/github/shoes4/shoes-core/lib/shoes/internal_app.rb:108
            initialize at /home/tobi/github/shoes4/shoes-core/lib/shoes/app.rb:49
                   app at /home/tobi/github/shoes4/shoes-core/lib/shoes/app.rb:19
                (root) at /home/tobi/github/shoes4/testing/para_fu.rb:1
                  load at org/jruby/RubyKernel.java:1087
                (root) at /home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:1
           execute_app at /home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:49
                   run at /home/tobi/github/shoes4/shoes-core/lib/shoes/ui/cli.rb:63
                  load at org/jruby/RubyKernel.java:1087
                (root) at /home/tobi/github/shoes4/bin/shoes-swt:5
```

**Comment by PragTob** [comment]
I'm at a loss as to how to fix this, I don't know how/why this works on other platforms :(

As it stands, `contents_alignment` doesn't run, therefore in `TextBlock::Painter#draw_layouts` `@dsl.gui.segments` is an empty array which leads to the crash. 

```ruby
        def draw_layouts(graphic_context)
          layouts = TextSegmentCollection.new(@dsl,
                                              @dsl.gui.segments,
                                              default_text_styles)
          layouts.paint_control(graphic_context)
        end
```

So it's either the guard we have in place right now, or a guard earlier in the process, or using `after_do` to let `contents_alignment` run before progress is initialized.

**Comment by jasonrclark** [comment]
Hmm, I'm confused too. If you're running straight on master or my diagnostic branch, how does the code ever get past https://github.com/shoes/shoes4/blob/master/shoes-swt/lib/shoes/swt/text_block/text_segment_collection.rb#L85 to crash where it's crashing?

FWIW, those are the same values that I get printed from running on OS X, but then it doesn't crash in `segment_ranges` because of line 85's guard.

**Comment by PragTob** [comment]
Oh yeah, sorry forgot to mention - I removed the guard for my fiddling, because for me the issue is all about removing the guard and making it work without it

**Comment by jasonrclark** [comment]
Ah, that totally makes sense with you removing the guard then! This leads me to two separate (but compatible) thoughts.

First off, I think that your precise case here would be addressed when we manage to fix #593 where we'd end up doing the equivalent of `contents_alignment` immediately after each element was added to its slot. In your test app, we add the `para`, then add the `progress` which triggers the `paint` where things fall over. After #593, we'd have done `contents_alignment` after adding the `para` so it'd have it's appropriate segments assigned.

Second thought--and what we might actually want to change here--is the question of whether `segment_ranges` is the appropriate location to be guarding on the segments being empty. With #1024 we absolutely can get an empty array, but we could probably gate up much higher, perhaps somewhere like  https://github.com/shoes/shoes4/blob/e2a8bfdf2eb6974e67e2f7c29a7612d023b52fcb/shoes-swt/lib/shoes/swt/text_block/painter.rb#L17 instead. Would you be happier with that placement @PragTob?

**Comment by PragTob** [comment]
You are so damn right about #593 - that's one of the big but potentially very hard things that's remaining, that's also HUGELY benefitial... maybe pre5? And then I'll take a good day on a weekend and hope it's enough :)

Straight in paintControl seems like a good place, I'll try that and PR :)
